File: py/56.合并区间.py
#
# @lc app=leetcode.cn id=56 lang=python3
#
# [56] 合并区间
#

# @lc code=start
class Solution:
    def merge(self, intervals):
        if intervals == []:
            return []
        intervals.sort(key = lambda x:x[0])
        re = []
        cur_arr = intervals[0]
        for i in intervals[1:]:
            
            if i[0] <= cur_arr[1]:
                cur_arr[0] = min(cur_arr[0], i[0])
                cur_arr[1] = max(cur_arr[1], i[1])
            else:
                re.append(cur_arr)
                cur_arr = i
        re.append(cur_arr)
        return re


# @lc code=end

i = [[1,3],[2,6],[8,10],[15,18]]
print(Solution().merge(i))

File: golang/56.合并区间.go
package main

import (
	"fmt"
	"sort"
)

/*
 * @lc app=leetcode.cn id=56 lang=golang
 *
 * [56] 合并区间
 */

// @lc code=start
// @lc code=start

func max(i, j int) int {
	if i >= j {
		return i
	}
	return j
}

func min(i, j int) int {
	if i <= j {
		return i
	}
	return j
}

func merge(intervals [][]int) [][]int {
	re := [][]int{}
	if len(intervals) == 0 {
		return re
	}
	// 排序
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	currentArr := []int{}
	currentArr = append(currentArr, intervals[0][0], intervals[0][1])

	for _, arr := range intervals[1:] {
		if arr[0] <= currentArr[1] {
			curMax := max(currentArr[1], arr[1])
			curMin := min(currentArr[0], arr[0])
			currentArr = append([]int{}, curMin, curMax)
		} else {
			re = append(re, currentArr)
			currentArr = append([]int{}, arr[0], arr[1])
		}
	}
	re = append(re, currentArr)

	return re
}

// @lc code=end
// @lc code=end

func main() {
	intervals := [][]int{}
	intervals = append(intervals, []int{1, 3}, []int{2, 6}, []int{8, 10}, []int{15, 18})
	fmt.Println(merge(intervals))
}
